plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBucketing.java

```diff
@@ -0,0 +1,219 @@
+        assertBucketAndHashEquals("decimal(38, 10)", "123456789012345.0", 93815101);
+
+        assertBucketAndHashEquals("binary", null, null);
+        assertBucketAndHashEquals("binary", ByteBuffer.wrap(new byte[]{}), 0);
+        assertBucketAndHashEquals("binary", ByteBuffer.wrap("hello trino".getBytes(StandardCharsets.UTF_8)), 493441885);
+        assertBucketAndHashEquals("binary", ByteBuffer.wrap("\uD843\uDFFC\uD843\uDFFD\uD843\uDFFE\uD843\uDFFF".getBytes(StandardCharsets.UTF_16)), 1291558121);
+    }
+
+    @Test(dataProvider = "unsupportedBucketingTypes")
+    public void testUnsupportedTypes(Type type)
+    {
+        assertThatThrownBy(() -> computeIcebergBucket(type, null, 1))
+                .hasMessage(format("Cannot bucket by type: %s", type));
+
+        assertThatThrownBy(() -> computeTrinoBucket(type, null, 1))
+                .hasMessage(format("Unsupported type for 'bucket': %s", toTrinoType(type, TYPE_MANAGER)));
+    }
+
+    @Test(dataProvider = "unsupportedTrinoBucketingTypes")
+    public void testUnsupportedTrinoTypes(Type type)
+    {
+        assertThatThrownBy(() -> computeTrinoBucket(type, null, 1))
+                .hasMessage(format("Cannot convert from Iceberg type '%s' (%s) to Trino type", type, type.typeId()));
+
+        assertNull(computeIcebergBucket(type, null, 1));
+    }
+
+    @DataProvider
+    public Object[][] unsupportedBucketingTypes()
+    {
+        return new Object[][] {
+            {BooleanType.get()},
+            {FloatType.get()},
+            {DoubleType.get()}};
+    }
+
+    @DataProvider
+    public Object[][] unsupportedTrinoBucketingTypes()
+    {
+        return new Object[][] {
+                {UUIDType.get()}};
+    }
+
+    private void assertBucketAndHashEquals(String icebergTypeName, Object value, Integer expectedHashValue)
+    {
+        Type icebergType = Types.fromPrimitiveString(icebergTypeName);
+        if (value != null && icebergType.typeId() == DECIMAL) {
+            value = new BigDecimal((String) value).setScale(((DecimalType) icebergType).scale());
+        }
+
+        assertBucketEquals(icebergType, value);
+        assertHashEquals(icebergType, value, expectedHashValue);
+    }
+
+    private void assertBucketEquals(Type icebergType, Object value)
+    {
+        assertBucketNumberEquals(icebergType, value, Integer.MAX_VALUE);
+
+        assertBucketNumberEquals(icebergType, value, 2);
+        assertBucketNumberEquals(icebergType, value, 7);
+        assertBucketNumberEquals(icebergType, value, 31);
+        assertBucketNumberEquals(icebergType, value, 32);
+        assertBucketNumberEquals(icebergType, value, 100);
+        assertBucketNumberEquals(icebergType, value, 10000);
+        assertBucketNumberEquals(icebergType, value, 524287); // prime number
+        assertBucketNumberEquals(icebergType, value, 1 << 30);
+    }
+
+    private void assertHashEquals(Type icebergType, Object value, Integer expectedHashValue)
```

Comment:
rename expectedHashValue -- this is not quite a hash value, right?

(in particular, you won't get a negative number here, while hash value can be negative)